Implement abm, bmi, bmi2, and tbm intrinsics

gnzlbg opened the pull request:
Implements all ABM, BMI, BMI2, and TBM intrinsics: 

- abm: POPCNT, LZCNT
- bmi: ANDN, BEXTR, BLSI ,BLSMSK, BLSR, TZCNT
- bmi2: BZHI , MULX, PDEP, PEXT
- tbm: BLCFILL, BLCI, BLCIC, BLCMSK, BLCS, BLSFILL, BLSIC, T1MSKC, TZMKS

This PR compiles the tests only for those architectures that support them. That is, to compile and run all the tests that a CPU can target one should use `RUSTFLAGS="-C target-cpu=native" cargo test`. 

Known issues:
- Add TBM's BEXTRI (the only instruction missing) which triggers an LLVM code-gen error #12 .

BurntSushi commented on src/x86/abm.rs:
```diff
@@ -0,0 +1,71 @@
+///
+/// When the operand is zero, it returns its size in bits.
+#[inline(always)]
+#[target_feature = "+sse4.2"]
```
Clang seems to use the `lzcnt` target feature: https://github.com/llvm-mirror/clang/blob/a38ba9770a70056f9fdd6c71f819e5db45a105e4/lib/Headers/lzcntintrin.h

Should we be doing the same?

gnzlbg replied:
@BurntSushi I have a comment on the top of the module with TODO mentioning this. IIRC  `lzcnt` is not a feature that `cfg(target_feature)` supports.  

That is, we would need to add support for `lzcnt` there, and in the future run-time detection support. The implementation cost of doing this is small, but all future code that wants to be portable would need to distinguish lzcnt from popcnt and SSE4.2. Whether this is worth it, I don't know, but I'd like to fill an issue in this repo to discuss it.

gnzlbg replied:
@BurntSushi I tried to find an alternative, but I couldn't, so I've fixed this.

BurntSushi commented on src/x86/abm.rs:
```diff
@@ -0,0 +1,71 @@
+
+/// Counts the bits that are set.
+#[inline(always)]
+#[target_feature = "+sse4.2"]
```
Clang uses the `popcnt` target feature: https://github.com/llvm-mirror/clang/blob/a38ba9770a70056f9fdd6c71f819e5db45a105e4/lib/Headers/popcntintrin.h

gnzlbg replied:
Same as for lzcnt.

BurntSushi commented on src/x86/bmi.rs:
```diff
@@ -0,0 +1,266 @@
+//! provides a quick overview of the available instructions.
+
+#[allow(dead_code)]
+extern "platform-intrinsic" {
```
Yeah so I'm explicitly trying not to use `platform-intrinsic` since I'd like to do away with that machinery. If you take a look at the other files, we link with the llvm intrinsics directly. My guess (based on [llvm intrinsic defs](https://gist.githubusercontent.com/anonymous/a25d3e3b4c14ee68d63bd1dcb0e1223c/raw/fb798be53a7b29f9b28a0c7a9e2b9a16ae915f71/allintrinsics) is that you probably want

```rust
extern {
    #[link_name = "llvm.x86.bmi.bextr.32"]
    fn bextr_u32(x: u32, y: u32) -> u32;
    #[link_name = "llvm.x86.bmi.bextr.64"]
    fn bextr_u32(x: u64, y: u64) -> u64;
}
```

gnzlbg replied:
@BurntSushi I did not knew this, should be fixed now.

BurntSushi replied:
Yeah technically we're still using `platform-intrinsic` for the cross platform simd llvm intrinsics (see `src/simd_llvm.rs`), but I'm trying to avoid them other than that. In particular, the platform intrinsics, IIRC, can't actually represent all of the vendor intrinsics. I think it's simpler to just write them out explicitly in Rust code instead of describing them in JSON.

BurntSushi commented on src/x86/bmi2.rs:
```diff
@@ -0,0 +1,202 @@
+    fn x86_bmi2_pdep_64(x: u64, y: u64) -> u64;
+    fn x86_bmi2_pext_32(x: u32, y: u32) -> u32;
+    fn x86_bmi2_pext_64(x: u64, y: u64) -> u64;
+}
```
Same as for bmi, let's be consistent and link to llvm intrinsics directly.

gnzlbg commented:
@BurntSushi I've moved the `platform-intrinsic`s to use `link_llvm_intrinsics`, and also added all of the TBM instruction set, except for `bextri`, which triggers a codegen error.

gnzlbg commented:
So I've checked the assembly of all generated intrinsics. It would be nice to set up travis-ci to do this automatically on each PR.

BurntSushi commented:
@gnzlbg All right this basically looks good to me! Is this ready to bring in? I will work on setting up the CI and polishing the tests.

gnzlbg commented:
@BurntSushi yes this is ready. I will implement the ARM intrinsics in a different branch.

BurntSushi commented:
Awesome thanks!
